File: core/src/main/scala/stasis/core/discovery/ServiceDiscoveryRequest.scala
package stasis.core.discovery

final case class ServiceDiscoveryRequest(
  isInitialRequest: Boolean,
  attributes: Map[String, String]
) {
  lazy val id: String = attributes.toList
    .sortBy(_._1)
    .map { case (k, v) => s"$k=$v" }
    .mkString("::")
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/discovery/ServiceDiscoveryRequest.kt
package stasis.client_android.lib.discovery

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class ServiceDiscoveryRequest(
    @field:Json(name = "is_initial_request")
    val isInitialRequest: Boolean,
    val attributes: Map<String, String>
) {
    val id: String by lazy {
        attributes.toList()
            .sortedBy { it.first }
            .joinToString(separator = "::") { (k, v) -> "$k=$v" }
    }
}
